internal/graph/storagewrapper.go

```diff
@@ -0,0 +1,162 @@
+package graph
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/cespare/xxhash/v2"
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+
+	"github.com/openfga/openfga/internal/keys"
+	"github.com/openfga/openfga/pkg/storage"
+	"github.com/openfga/openfga/pkg/tuple"
+)
+
+var _ storage.OpenFGADatastore = (*cachedDatastore)(nil)
+
+type cachedDatastore struct {
+	storage.OpenFGADatastore
+
+	cache        storage.InMemoryCache[any]
+	maxCacheSize int64
+	ttl          time.Duration
+}
+
+// NewCachedDatastore returns a wrapper over a datastore...
+func NewCachedDatastore(inner storage.OpenFGADatastore, cache storage.InMemoryCache[any], maxSize int64, ttl time.Duration) *cachedDatastore {
+	return &cachedDatastore{
+		OpenFGADatastore: inner,
+		cache:            cache,
+		maxCacheSize:     maxSize,
+		ttl:              ttl,
+	}
+}
+
+func (c *cachedDatastore) ReadUsersetTuples(ctx context.Context, store string, filter storage.ReadUsersetTuplesFilter, options storage.ReadUsersetTuplesOptions) (storage.TupleIterator, error) {
+	iter, err := c.OpenFGADatastore.ReadUsersetTuples(ctx, store, filter, options)
+	if err != nil {
+		return nil, err
+	}
+
+	cacheKey := fmt.Sprintf("%s/%s#%s", store, filter.Object, filter.Relation)
+	// todo(jpadilla): stable sort filter.AllowedUserTypeRestrictions
+	for _, userset := range filter.AllowedUserTypeRestrictions {
+		if _, ok := userset.GetRelationOrWildcard().(*openfgav1.RelationReference_Relation); ok {
+			cacheKey += "/" + userset.GetType() + ":%#" + userset.GetRelation()
+		}
+		if _, ok := userset.GetRelationOrWildcard().(*openfgav1.RelationReference_Wildcard); ok {
+			cacheKey += "/" + userset.GetType() + ":*"
+		}
+	}
+
+	return c.newCachedIterator(iter, cacheKey), nil
+}
+
+func (c *cachedDatastore) Read(ctx context.Context, store string, tupleKey *openfgav1.TupleKey, options storage.ReadOptions) (storage.TupleIterator, error) {
+	iter, err := c.OpenFGADatastore.Read(ctx, store, tupleKey, options)
+	if err != nil {
+		return nil, err
+	}
+	cacheKey := fmt.Sprintf("%s/%s", store, tuple.TupleKeyToString(tupleKey))
+	return c.newCachedIterator(iter, cacheKey), nil
+}
+
+func (c *cachedDatastore) newCachedIterator(iter storage.TupleIterator, key string) storage.TupleIterator {
+	cacheKey := cacheKeyFor(key)
+	cachedResp := c.cache.Get(cacheKey)
+	isCached := cachedResp != nil && !cachedResp.Expired && cachedResp.Value != nil
+	if isCached {
+		return storage.NewStaticTupleIterator(cachedResp.Value.([]*openfgav1.Tuple))
+	}
+
+	return &cachedIterator{
+		iter:         iter,
+		tks:          make([]*openfgav1.Tuple, 0, c.maxCacheSize),
```

Comment:
I would suggest setting the length of the slice equal to `maxCacheSize` and using a variable to track the current index position. Accessing a slice by index is far more efficient than using `append()` to populate a slice of known length.

This will give your type better compute performance, and will require the same amount of memory. (setting the capacity allocates a backing array of the given size).

Reply:
Ufff we can apply this optimization pretty much everywhere in the codebase then 😅

Reply:
@senojj ah the only downside i see with that is either we possibly cache partially populated slices or _waste_ cycles filtering out all the nil values. `maxCacheSize` is just the possible ceiling, won't know the actual size of the slice at this point.

Reply:
You would know the the current position in the slice because you would keep track of it with a variable.
```go
var position int
```
You wouldn't need to filter out any `nil` values because you would just slice the slice when you flush.
```go
c.tks[:position]
```

Reply:
Also note that you already enforce that you only ever cache up to `maxCacheSize` in the `Next()` method:
```go
if len(c.tks) < int(c.maxCacheSize) {
	c.tks = append(c.tks, tuple)
}
```
Therefore, at this time it is an enforced ceiling, not just a possible ceiling. However, I do also question if this could result in possible partial results being cached.